logstash-core/src/main/java/org/logstash/instrument/witness/schedule/WitnessScheduler.java

```diff
@@ -0,0 +1,82 @@
+package org.logstash.instrument.witness.schedule;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Schedules {@link ScheduledWitness} to refresh themselves on an interval.
+ */
+public class WitnessScheduler {
+
+    private final ScheduledWitness witness;
+    private final ScheduledExecutorService executorService;
+    private static final Logger LOGGER = LogManager.getLogger(WitnessScheduler.class);
+
+    /**
+     * Constructor
+     *
+     * @param witness the {@link ScheduledWitness} to schedule
+     */
+    public WitnessScheduler(ScheduledWitness witness) {
+        this.witness = witness;
+
+        this.executorService = Executors.newScheduledThreadPool(1, ((Runnable r) -> {
+            Thread t = new Thread(r);
+            //Allow this thread to simply die when the JVM dies
+            t.setDaemon(true);
+            //Set the name
+            t.setName(witness.threadName());
+            //Set the priority
+            t.setPriority(witness.priority());
+            return t;
+        }));
+    }
+
+    /**
+     * Schedules the witness to refresh on provided schedule. Note - this implementation does not allow refreshes faster then every 1 second.
+     */
+    public void schedule() {
+        executorService.scheduleAtFixedRate(new RefreshRunnable(), 0, witness.every().getSeconds(), TimeUnit.SECONDS);
+    }
+
+    /**
+     * Shuts down the underlying executor service. Since these are daemon threads, this is not absolutely necessary.
+     */
+    public void shutdown(){
+        executorService.shutdown();
+        try {
+            if(!executorService.awaitTermination(5, TimeUnit.SECONDS)){
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            //do nothing
```

Comment:
@jakelandis are we sure we want to suppress an interrupt here? (won't this interfere with orderly shutdowns on a single outside interrupt? ... just asking, I'm not fully sure what Thread this will be called from)

Reply:
Shutdown will only be called from the main thread, if called at all. The executor service will always contain daemon threads, so on JVM termination there is no need to call this method. Honestly this method really only exists to help testing.

Reply:
@jakelandis 

> Shutdown will only be called from the main thread, if called at all

but **if** called from the main thread it will remove its interrupt status? Which may be a problem?

Reply:
How/why would anyone signal interrupt to the main thread ?  It depends on what we define as the interrupt policy for main thread.  IMO, the main thread should be a `do nothing` policy.

WDYT?

Reply:
@jakelandis argh so sorry missed answering this one before you pushed changes :(
That said:

> How/why would anyone signal interrupt to the main thread?

Exactly, I don't know :D and it probably shouldn't happen here ever. => rethrow `IllegalstateException` imo :)

Reply:
@original-brownbear - sounds good.  commit here : https://github.com/elastic/logstash/pull/8091/commits/f75c1987edaba437117af9d92224f564fea5766b